### PR TITLE
Translate IN predicate to connector expression - continue

### PR DESCRIPTION
**Pull request by assaf2**
## Description

Adds translation for InPredicate and InListExpression to connector expression form.

> Is this change a fix, improvement, new feature, refactoring, or other?

New feature.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core engine & postgres connector.

> How would you describe this change to a non-technical end user or system administrator?

This enabled pushdown of IN predicates which are not expressible with Domain (i.e. references other symbols or contains function calls)

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# PostgreSQL connector
* Improve performance of queries involving `IN` expression by pushing predicate computation to the PostgreSQL database. (`#13136`)
```

### CODE REVIEW

**Comment by assaf2** [comment]
I've asked @wendigo to take over his PR https://github.com/trinodb/trino/pull/11396.
I've dropped "Rewrite typed NULL char/varchar constants" as @findepi suggested.
I've also dropped "Add explicit cast pushdown support to PostgreSQL" since it now seems to me non-mandatory to the PR and https://github.com/trinodb/trino/pull/11396#discussion_r867732779 is a significant problem. I believe there might be more problematic use cases so IMO should be done on a separate PR.
I've addressed https://github.com/trinodb/trino/pull/11396#discussion_r867761861

**Comment by findepi** [comment]
> @assaf2 @wendigo is it supposed to supersede #11396?

thanks for 👍  on that question, @wendigo .

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteVarcharConstant.java [hunk 1]
```diff
@@ -40,6 +40,9 @@ public Pattern<Constant> getPattern()
     public Optional<String> rewrite(Constant constant, Captures captures, RewriteContext<String> context)
     {
         Slice slice = (Slice) constant.getValue();
+        if (slice == null) {
+            return Optional.empty();
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteVarcharConstant.java:44]: per https://github.com/trinodb/trino/pull/11396/commits/6c89d7c752d8839e94dab3ff2ec534fd55b6b904#r867723626
at least `RewriteExactNumericConstant` also needs such change (maybe others too)

**Review by findepi — APPROVED** [review]

**Comment by assaf2** [comment]
> @assaf2 @wendigo is it supposed to supersede #11396?

IMO yes, the dropped commits:

- "Rewrite typed NULL char/varchar constants"
- "Add explicit cast pushdown support to PostgreSQL" 

can be fixed and re-added on a separated PR, if @wendigo thinks they are still necessary

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteVarcharConstant.java (hunk 1 above)
Comment (assaf2) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteVarcharConstant.java:44]: Added, I don't see any other constant rewriters at `io.trino.plugin.jdbc.expression` so I assume these are the only places

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
thank you @wendigo @assaf2

**Pull request merged**